tempel.el

```diff
@@ -220,11 +225,25 @@ REGION are the current region bounds."
           (tempel--inhibit-hooks t))
       (tempel--disable (overlay-get ov 'tempel--range)))))
 
+(defun tempel--ensure-ov-invariant (st ov)
+  "Ensure the overlay order is maintained, by clamping the boundaries
+of every overlay which was added to the state ST before OV to the
+start of OV, and likewise for overlays added to ST after OV."
+  (let ((start (overlay-start ov))
+        (end (overlay-end ov))
+        before-current)
+    (dolist (ov_ (cdar st))
+      (if (eq ov ov_) (setq before-current t)
+        (if before-current
+            (move-overlay ov_ (min start (overlay-start ov_)) (min start (overlay-end ov_)))
+          (move-overlay ov_ (max end (overlay-start ov_)) (max end (overlay-end ov_))))))))
```

Comment:
I'm pretty sure it should be safe to just enforce this on the overlays _after_ the current overlay (so we could stop the loop when `(eq ov ov_)`). But I'm not completely sure, do you have opinions?

Reply:
Proof: we start in a state where the order of overlay markers is right. Markers with different positions or with the same insertion type never change their relative order after a buffer edit. If they have the same position and different insertion type, the one with insertion type `t` can move forward and the other stay fixed.

The only field or form overlay marker with insertion type `t` is the marker at the start of a form overlay. If it moves forward relative to a marker of other overlay, this marker should be the end marker or be at the same position of the end marker since the state began ordered. But then the state is still ordered.

The other way for markers to change position is if we set the position in code. But in the code we are only moving the end of form/field overlays to the right (forward direction), so it suffices to enforce the invariant to the right.

Reply:
Thanks, my opinion was that there should be a proof. ;)

Please continue improving the code in your PRs - maybe you find better/more robust solutions. I will definitely merge these fixes. I just want to let you know that I won't look often at your PR due to time constraints - so please don't worry when I am not responding soon. Right now we still have to wait for the FSF assignment anyway.